Signal/src/call/PeerConnectionClient.swift

```diff
@@ -70,12 +70,24 @@ class PeerConnectionClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelD
              dataChannelSignaling = "signaling"
     }
 
+    // A state in this class should only be accessed on this queue in order to
+    // serialize access.
+    //
+    // This queue is also used to perform expensive calls to the WebRTC API.
+    private static let signalingQueue = DispatchQueue(label: "CallServiceSignalingQueue")
+
     // Delegate is notified of key events in the call lifecycle.
-    public weak var delegate: PeerConnectionClientDelegate!
+    private weak var delegate: PeerConnectionClientDelegate!
+
+    func setDelegate(delegate: PeerConnectionClientDelegate?) {
+        PeerConnectionClient.signalingQueue.async {
+            self.delegate = delegate
+        }
```

Comment:
Nearly all of the signalling-related concurrency is now async, reducing the surface area for deadlock.